<a href="https://colab.research.google.com/github/AlexParreira/AlexParreira_Data_Science/blob/master/Pokemon_Train_Machine_learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pokemon


<sup>Origem: Wikipédia, a enciclopédia livre</sup>
<br>
Pokémon é uma série de jogos eletrônicos desenvolvidos pela Game Freak e publicados pela Nintendo como parte da franquia de mídia Pokémon. Lançado pela primeira vez em 1996 no Japão para o console Game Boy, a principal série de jogos de RPGs, que continuou em cada geração em portáteis da Nintendo.

# Os Dados


O conjunto de dados contém 13 atributos com 800 amostras.
<br>
<sub><a href='https://www.kaggle.com/abcsds/pokemon>'>Fonte do csv</a>
<br>

Número do índice PokeDex
Nome
Tipo 1 - Tipo de pokemon
Tipo 2 - Tipo secundário de pokemon
HP - Pontos de vida
Ataque - Ponto de ataque
Defesa - ponto de defesa
Sp. Atk - Ponto de ataque especial
Sp. Def - ponto defensivo especial
Rapidez
Total - Soma do Ataque, Sp. Atk, Defesa, Sp. Def, Velocidade e HP
Geração
Lendário - Se o pokemon é lendário ou não

# Pré processamento

In [371]:
# montando um drive no Colab
from google.colab import drive

drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [0]:
# Importando os pacotes necessários
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Permite a visualização de todas as colunas
pd.options.display.max_columns = None

# Renderiza os gráficos direto no notebook
%matplotlib inline

In [373]:
df = pd.read_csv('/content/drive/My Drive/data/Pokemon/Pokemon.csv')

df.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [374]:
# Verifica as Dimensões do DataFrame
print('Variáveis:\t{}\nEntradas:\t{}'.format(df.shape[1], df.shape[0]))

Variáveis:	13
Entradas:	800


O conjunto de dados contém 13 atributos com 800 amostras e o meu objetivo será descobrir se um determinado pokemon é lendário ou não.

In [375]:
print(df.apply(lambda x: x.count()))

#             800
Name          800
Type 1        800
Type 2        414
Total         800
HP            800
Attack        800
Defense       800
Sp. Atk       800
Sp. Def       800
Speed         800
Generation    800
Legendary     800
dtype: int64


O Type 2 e a unica coluna faltante pois nem todos pokemons possuem 2 tipos por tando, repetirei seu tipo primario para o tipo 2.

In [0]:
df['Type 2'].fillna(df['Type 1'], inplace = True)

In [377]:
# Ver a porcentagem valores faltantes
(df.isnull().sum()/df.shape[0]).sort_values(ascending=False)

Legendary     0.0
Generation    0.0
Speed         0.0
Sp. Def       0.0
Sp. Atk       0.0
Defense       0.0
Attack        0.0
HP            0.0
Total         0.0
Type 2        0.0
Type 1        0.0
Name          0.0
#             0.0
dtype: float64

In [378]:
display(df.dtypes)
display(df.head())

#              int64
Name          object
Type 1        object
Type 2        object
Total          int64
HP             int64
Attack         int64
Defense        int64
Sp. Atk        int64
Sp. Def        int64
Speed          int64
Generation     int64
Legendary       bool
dtype: object

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire    Fire    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [0]:
# A variável Name não e compativel com os teste
df = df.drop(['Name'], axis = 1)

A variável Type 1 e Type 2 precisa ter seus valores convertidos para númericos.

In [380]:
# criando valores únicos para o Type 1
i = 0
uniqueItem = dict()

for item in df['Type 1']:
  if item not in uniqueItem:
    uniqueItem[str(item)] = i
    i+=1
uniqueItem

{'Bug': 3,
 'Dark': 15,
 'Dragon': 14,
 'Electric': 6,
 'Fairy': 8,
 'Fighting': 9,
 'Fire': 1,
 'Flying': 17,
 'Ghost': 12,
 'Grass': 0,
 'Ground': 7,
 'Ice': 13,
 'Normal': 4,
 'Poison': 5,
 'Psychic': 10,
 'Rock': 11,
 'Steel': 16,
 'Water': 2}

18 tipos diferentes foram gerados.

In [0]:
# inserindo o uniqueItem do DataFrame
for type in df['Type 1']:
  df = df.replace({type:uniqueItem.get(type)})

In [382]:
df.head()

#  Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
0  1       0       5    318  45      49       49       65       65     45   
1  2       0       5    405  60      62       63       80       80     60   
2  3       0       5    525  80      82       83      100      100     80   
3  3       0       5    625  80     100      123      122      120     80   
4  4       1       1    309  39      52       43       60       50     65   

   Generation  Legendary  
0           1      False  
1           1      False  
2           1      False  
3           1      False  
4           1      False

In [0]:
# dividindo dados de teste e treino
x = np.array(df.iloc[:,0:-1])
y = np.array([[df['Legendary']]])

In [0]:
y = y.reshape(800)

# Construindo modelo e avaliação

In [0]:
# importando as bibliotecas do modelo de Machine learning
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn import linear_model

In [386]:
#modelo Gaussian Naive Bayes

X_train, X_test, y_train, y_test = train_test_split(
x, y, test_size = 0.2, random_state = 10)

model = GaussianNB()
model.fit(X_train, y_train)


GaussianNB(priors=None, var_smoothing=1e-09)

In [387]:
# Precisão do teste do modelo Gaussian Naive Bayes
y_pred = model.predict(X_test)
print("Accuracy score of NaiveBayes:",accuracy_score(y_test,y_pred))

Accuracy score of NaiveBayes: 0.94375


In [389]:
# Precisão do teste SDG Classifier
clf = linear_model.SGDClassifier()
clf.fit(x, y)

y_pred = clf.predict(X_test)
print("Accuracy score of SDG: ",accuracy_score(y_test,y_pred))

Accuracy score of SDG:  0.8875


# Conclusão


Neste conjunto de dados, tentei implementar os métodos de aprendizado de máquina, esse conjunto de dados me ensinou muitas coisas. Especialmente na parte do pré-processamento, teve que mergulhar nas características dos pandas. Graças ao esquema no site do sci-kit, implementei um novo classificador.
Como um resultado,
* Pontuação de precisão de Naive Bayes: 0.94375
* Pontuação de precisão de ODS: 0.8875